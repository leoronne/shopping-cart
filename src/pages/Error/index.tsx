import React  from 'react';

import { Container, Title } from './styles';
import error from '~/assets/img/error.png';

const Error: React.FC = () => {
  return (
    <Container>
    <img src={error} alt="Error" className="error-bg"/>
      <Title>
          Oops...
      </Title>
      <p>Sorry, we had a problem loading our resources.</p>
      <p>Please, try again later!</p>
    </Container>
  );
};

export default Error;
