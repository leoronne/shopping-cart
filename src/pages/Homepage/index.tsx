import React, { useRef, useContext, useState } from 'react';
import ReactTooltip from 'react-tooltip';

import DotLoaderComp from '../../components/DotLoaderComp';

import notify from '../../services/toast';
import formatNumber from '../../utils/formatNumber';

import { Context as ProductContext } from '../../Context/ProductsContext';
import { Context as VoucherContext } from '../../Context/VouchersContext';

import banana from '~assets/svg/banana.svg';
import apple from '~assets/svg/apple.svg';
import mango from '~assets/svg/mango.svg';
import orange from '~assets/svg/orange.svg';
import noicon from '~assets/svg/noicon.svg';

import {
  Container,
  GridContainer,
  ProductsContainer,
  ProductsListcontainer,
  ProductsCardcontainer,
  ProductsCardInfo,
  ProductsCardValues,
  BuyButton,
  ShoppingCartContainer,
  ShoppingCart,
  ProductList,
  ProductRowContent,
  CartProductIcon,
  ProductIcon,
  PlusMinusContainer,
  MinusIcon,
  PlusIcon,
  DiscountCode,
  TotalValues,
  RowValues,
} from './styles';

export interface ProductItemProps {
  quant?: number;
}

export interface BuyButtonProps {
  disabled?: boolean;
}

export interface CartProps {
  icon?: string;
}

const Homepage: React.FC = () => {
  const {
    products,
    setCartItems,
    cartItems,
    addProducts,
    removeProducts,
    totalItemValue,
    shipping,
    totalValue,
    discount,
    setDiscount,
    voucherApplied,
    setVoucherApplied,
    voucherCode2,
    setVoucherCode2,
    discountValue,
    setDiscountValue,
  } = useContext(ProductContext);
  const { vouchers } = useContext(VoucherContext);
  const [voucherCode, setVoucherCode] = useState('');
  const [loadingOrder, setLoadingOrder] = useState(false);
  const productsRef = useRef() as React.MutableRefObject<HTMLDivElement>;

  const handleSubmit = () => {
    ReactTooltip.hide();
    setLoadingOrder(true);
    setTimeout(() => {
      notify('Thank you! We received your order!', 'success');
      setCartItems([]);
      setLoadingOrder(false);
    }, 1000);
  };

  const removeDiscounts = () => {
    setDiscount({ type: null, value: 0, desc: '' });
    setDiscountValue(0);
    setVoucherApplied(false);
    setVoucherCode('');
  };

  const handleDiscounts = () => {
    ReactTooltip.hide();
    if (voucherCode) {
      if (voucherApplied) notify('Sorry, you have already applied a voucher', 'error');
      const cupom = vouchers.find((voucher) => voucher.code === voucherCode);

      if (!cupom) {
        setVoucherCode('');
        setDiscount({ type: null, value: 0, desc: '' });
        setDiscountValue(0);
        return notify('Sorry, you entered an invalid voucher', 'error');
      }
      setDiscount({
        type: cupom.type,
        value: cupom.type === 'shipping' ? cupom.minValue : cupom.amount,
        desc: `The Voucher qualifies for a ${cupom.type} discount`,
      });

      notify('Voucher Applied!', 'success');
      setVoucherApplied(true);
      setVoucherCode2(voucherCode);
    }
  };

  function returnIcon(name: string) {
    switch (name) {
      case 'Banana':
        return banana;
      case 'Apple':
        return apple;
      case 'Orange':
        return orange;
      case 'Mango':
        return mango;
      default:
        return noicon;
    }
  }

  function ProductCard(props: { name: string; value: number; quant: number; id: number }) {
    return (
      <ProductsCardcontainer quant={props.quant}>
        <div className="cart-product-icon">
          <ProductIcon icon={returnIcon(props.name)} />
        </div>
        <ProductsCardInfo>
          <div className="text--center">
            <strong>{props.name}</strong>
          </div>
          <ProductsCardValues>
            <span>{formatNumber(props.value)}</span>
            <br />
            <span>{`${props.quant} left`}</span>
          </ProductsCardValues>
        </ProductsCardInfo>

        <BuyButton
          onClick={() => {
            if (props.quant > 0) addProducts(props.id);
          }}
          disabled={props.quant > 0 ? false : true}
          data-tip={props.quant > 0 ? 'Add item to cart' : `There is no more of this item on stock`}
        >
          BUY
        </BuyButton>
      </ProductsCardcontainer>
    );
  }

  function ProductRow(props: { name: string; value: number; quant: number; id: number }) {
    return (
      <ProductRowContent>
        <CartProductIcon>
          <ProductIcon icon={returnIcon(props.name)} />
        </CartProductIcon>
        <div className="product-info">
          <div className="product-name">
            <span>{props.name}</span>
          </div>
          <div className="product-qnt-values">
            <span>Quantity: {props.quant}</span>
            <span>{formatNumber(props.value * props.quant)}</span>
          </div>
        </div>
        <PlusMinusContainer>
          <span>
            <PlusIcon
              onClick={() => {
                if (props.quant > 0) addProducts(props.id);
              }}
            />
          </span>
          <span>
            <MinusIcon disabled={false} onClick={() => removeProducts(props.id)} />
          </span>
        </PlusMinusContainer>
      </ProductRowContent>
    );
  }

  function ValueRow(props: { name: string; value: number; isTotal: Boolean }) {
    const isDiscount = props.name === 'Discount' && discountValue > 0;
    return (
      <div className={`cart-row ${props.isTotal ? ' total' : ''}`} data-tip={props.name === 'Discount' ? discount.desc : null}>
        <p>{props.name}</p>
        <RowValues>
          {isDiscount ? (
            <div className="minus-icon">
              <MinusIcon disabled={true} />
            </div>
          ) : (
            ''
          )}
          <p className={isDiscount ? 'discount' : ''}>{formatNumber(props.value)}</p>
        </RowValues>
      </div>
    );
  }
  // <ReactTooltip place="left" type="dark" effect="solid" />
  return (
    <Container>
      <GridContainer>
        <ProductsContainer>
          <ProductsListcontainer>
            {products && products.length === 0
              ? ''
              : products.map((product) => (
                  <ProductCard name={product.name} value={product.price} quant={product.available} id={product.id} key={product.id} />
                ))}
          </ProductsListcontainer>
        </ProductsContainer>

        <ShoppingCartContainer>
          <ShoppingCart>
            <div className="title">
              <p>Shopping Cart</p>
            </div>
            <div className="cart-container">
              <ProductList ref={productsRef}>
                {cartItems.length === 0
                  ? ''
                  : cartItems.map((item) => {
                      if (item.available > 0)
                        return <ProductRow name={item.name} value={item.price} quant={item.available} id={item.id} key={item.id} />;
                      return '';
                    })}
              </ProductList>

              <DiscountCode>
                {voucherApplied ? (
                  <>
                    <input type="text" value={voucherCode2} placeholder="Voucher Code" name="voucher" disabled={true} />
                    <button type="button" data-tip="Remove Voucher" onClick={removeDiscounts}>
                      REMOVE
                    </button>
                  </>
                ) : (
                  <>
                    <input
                      type="text"
                      value={voucherCode}
                      placeholder="Voucher Code"
                      name="voucher"
                      onChange={(e) => {
                        setVoucherCode(e.target.value);
                      }}
                    />
                    <button type="button" data-tip="Apply Voucher" onClick={handleDiscounts} disabled={!voucherCode ? true : false}>
                      APPLY
                    </button>
                  </>
                )}
              </DiscountCode>
            </div>

            <TotalValues>
              <ValueRow name="Subtotal" value={totalItemValue} isTotal={false} />
              <ValueRow name="Shipping" value={cartItems.length === 0 ? 0 : shipping} isTotal={false} />
              <ValueRow name="Discount" value={discountValue} isTotal={false} />
              <ValueRow name="Total" value={cartItems.length === 0 ? 0 : totalValue} isTotal={true} />
            </TotalValues>
          </ShoppingCart>
          <button type="button" disabled={cartItems.length === 0 || loadingOrder ? true : false} className="checkout-button" onClick={handleSubmit} data-tip="Submit Order">
            <DotLoaderComp loading={loadingOrder} size={15} color="#fff" defaultText="CHECKOUT" />
          </button>
        </ShoppingCartContainer>
      </GridContainer>
      <ReactTooltip place="bottom" type="dark" effect="solid" />
    </Container>
  );
};

export default Homepage;
