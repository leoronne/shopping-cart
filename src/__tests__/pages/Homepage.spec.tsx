import React from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react';

import AppProvider from '../../hooks';

import { Homepage } from '../../pages';

describe('Homepage', () => {
  it('should render', async () => {
    await act(async () => {
      const { asFragment } = render(
        <AppProvider>
          <Homepage />
        </AppProvider>
      );

      const html = asFragment();

      expect(html).toMatchSnapshot();
    });
  });

  it('should be able to apply vouchers', async () => {
    await act(async () => {
      const { getByTestId } = render(
        <AppProvider>
          <Homepage />
        </AppProvider>
      );
      const voucherInput = (await waitFor(() => getByTestId('voucher-input'))) as HTMLInputElement;

      const value = '#100DOLLARS';

      fireEvent.change(voucherInput, { target: { value } });

      expect(voucherInput.value).toBe(value);

      const submitButton = await waitFor(() => getByTestId('submit-voucher'));

      fireEvent.click(submitButton);

      const discountValue = await waitFor(() => getByTestId('Discount'));

      await waitFor(() => {
        expect(discountValue.textContent).toContain('100.00');
      });
    });
  });

  it('should be able to remove vouchers', async () => {
    await act(async () => {
      const { getByTestId } = render(
        <AppProvider>
          <Homepage />
        </AppProvider>
      );
      let voucherInput = (await waitFor(() => getByTestId('voucher-input'))) as HTMLInputElement;

      const value = '#100DOLLARS';

      fireEvent.change(voucherInput, { target: { value } });

      expect(voucherInput.value).toBe(value);

      const submitButton = await waitFor(() => getByTestId('submit-voucher'));

      fireEvent.click(submitButton);

      const discountValue = await waitFor(() => getByTestId('Discount'));

      expect(discountValue.textContent).toContain('100.00');

      const removeButton = await waitFor(() => getByTestId('remove-voucher'));

      fireEvent.click(removeButton);

      voucherInput = (await waitFor(() => getByTestId('voucher-input'))) as HTMLInputElement;

      await waitFor(() => {
        expect(voucherInput.value).toBe('');
        expect(discountValue.textContent).not.toContain('100.00');
      });
    });
  });

  it('should not be able to apply wrong vouchers', async () => {
    await act(async () => {
      const { getByTestId } = render(
        <AppProvider>
          <Homepage />
        </AppProvider>
      );
      const voucherInput = (await waitFor(() => getByTestId('voucher-input'))) as HTMLInputElement;

      const value = '#WRONGCODE';

      fireEvent.change(voucherInput, { target: { value } });

      expect(voucherInput.value).toBe(value);

      const submitButton = await waitFor(() => getByTestId('submit-voucher'));

      fireEvent.click(submitButton);

      const discountValue = await waitFor(() => getByTestId('Discount'));

      await waitFor(() => {
        expect(voucherInput.value).toBe('');
        expect(discountValue.textContent).toContain('0.00');
      });
    });
  });

  it('should be able to add products to cart', async () => {
    await act(async () => {
      const { getByTestId, queryByTestId } = render(
        <AppProvider>
          <Homepage />
        </AppProvider>
      );
      // Product 1
      const quantityProduct1 = await waitFor(() => getByTestId('quantity-left-product-1'));

      // Original quantity
      const quantityLeftProduct1 = quantityProduct1.textContent.replace(' left', '');

      const buyProduct1 = await waitFor(() => getByTestId('buy-button-product-1'));

      fireEvent.click(buyProduct1);

      // Updated Quantity
      expect(quantityProduct1.textContent).toBe(`${Number(quantityLeftProduct1) - 1} left`);

      // Product 2
      const quantityProduct2 = await waitFor(() => getByTestId('quantity-left-product-2'));

      // Original quantity
      const quantityLeftProduct2 = quantityProduct2.textContent.replace(' left', '');

      const buyProduct2 = await waitFor(() => getByTestId('buy-button-product-2'));

      fireEvent.click(buyProduct2);

      // Updated Quantity
      expect(quantityProduct2.textContent).toBe(`${Number(quantityLeftProduct2) - 1} left`);

      // Check if it is on the cart
      await waitFor(() => {
        expect(queryByTestId('cart-item-1')).toBeTruthy();
        expect(queryByTestId('cart-item-2')).toBeTruthy();
      });
    });
  });

  it('should be able to alter the quantity of the items on the cart', async () => {
    await act(async () => {
      const { getByTestId, queryByTestId } = render(
        <AppProvider>
          <Homepage />
        </AppProvider>
      );
      // Product 1
      const quantityProduct1 = await waitFor(() => getByTestId('quantity-left-product-1'));

      // Original quantity
      const quantityLeftProduct1 = quantityProduct1.textContent.replace(' left', '');

      const buyProduct1 = await waitFor(() => getByTestId('buy-button-product-1'));

      fireEvent.click(buyProduct1);

      // Updated Quantity
      expect(quantityProduct1.textContent).toBe(`${Number(quantityLeftProduct1) - 1} left`);
      expect(queryByTestId('cart-item-1')).toBeTruthy();

      // Add one more
      const addProduct1 = await waitFor(() => getByTestId('add-cart-item-1'));

      fireEvent.click(addProduct1);

      const quantityItem1 = await waitFor(() => getByTestId('cart-item-1-quantity'));

      expect(quantityItem1.textContent).toBe(`Quantity: 2`);
      expect(quantityProduct1.textContent).toBe(`${Number(quantityLeftProduct1) - 2} left`);

      // Remove all
      const removeProduct1 = await waitFor(() => getByTestId('remove-cart-item-1'));

      fireEvent.click(removeProduct1);
      fireEvent.click(removeProduct1);

      expect(queryByTestId('cart-item-1')).not.toBeTruthy();
      expect(quantityProduct1.textContent).toBe(`${quantityLeftProduct1} left`);
    });
  });

  it('should be able to submit order', async () => {
    await act(async () => {
      const { getByTestId, queryByTestId } = render(
        <AppProvider>
          <Homepage />
        </AppProvider>
      );
      // Product 1
      const quantityProduct1 = await waitFor(() => getByTestId('quantity-left-product-1'));

      // Original quantity
      const quantityLeftProduct1 = quantityProduct1.textContent.replace(' left', '');

      const buyProduct1 = await waitFor(() => getByTestId('buy-button-product-1'));

      fireEvent.click(buyProduct1);

      // Updated Quantity
      expect(quantityProduct1.textContent).toBe(`${Number(quantityLeftProduct1) - 1} left`);

      // Product 2
      const quantityProduct2 = await waitFor(() => getByTestId('quantity-left-product-2'));

      // Original quantity
      const quantityLeftProduct2 = quantityProduct2.textContent.replace(' left', '');

      const buyProduct2 = await waitFor(() => getByTestId('buy-button-product-2'));

      fireEvent.click(buyProduct2);

      // Updated Quantity
      expect(quantityProduct2.textContent).toBe(`${Number(quantityLeftProduct2) - 1} left`);

      // Check if it is on the cart
      expect(queryByTestId('cart-item-1')).toBeTruthy();
      expect(queryByTestId('cart-item-2')).toBeTruthy();

      const submitButton = await waitFor(() => getByTestId('submit-order'));

      fireEvent.click(submitButton);

      await waitFor(() => {
        expect(queryByTestId('cart-item-1')).not.toBeTruthy();
        expect(queryByTestId('cart-item-2')).not.toBeTruthy();
      });
    });
  });
});
