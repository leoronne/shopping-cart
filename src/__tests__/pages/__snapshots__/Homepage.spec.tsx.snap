// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage should render 1`] = `
<DocumentFragment>
  <div
    class="sc-fKFyDc fCoja"
    data-testid="product-cointainer"
  >
    <div
      class="sc-bBXqnf gCbUiR"
    >
      <div
        class="sc-iqHYGH gDIMJP"
      >
        <div
          class="sc-crrsfI Wezfb"
        />
      </div>
      <div
        class="sc-dIUggk kofTMZ"
      >
        <div
          class="sc-hHftDr fIpfqg"
        >
          <div
            class="title"
          >
            <p>
              Cart
            </p>
          </div>
          <div
            class="cart-container"
          >
            <div
              class="sc-dmlrTW gukuJa"
            />
            <div
              class="sc-kfzAmx dIygDd"
            >
              <form>
                <input
                  data-testid="voucher-input"
                  name="voucher"
                  placeholder="Voucher Code"
                  type="text"
                  value=""
                />
                <button
                  data-testid="submit-voucher"
                  disabled=""
                  type="submit"
                >
                  APPLY
                </button>
              </form>
            </div>
          </div>
          <div
            class="sc-dQppl eJWKvx"
          >
            <div
              class="cart-row "
            >
              <p>
                Subtotal
              </p>
              <div
                class="sc-bqyKva pvONl"
              >
                <p
                  class="Subtotal"
                  data-testid="Subtotal"
                >
                  USD 0.00
                </p>
              </div>
            </div>
            <div
              class="cart-row "
            >
              <p>
                Shipping
              </p>
              <div
                class="sc-bqyKva pvONl"
              >
                <p
                  class="Shipping"
                  data-testid="Shipping"
                >
                  USD 0.00
                </p>
              </div>
            </div>
            <div
              class="cart-row "
            >
              <p>
                Discount
              </p>
              <div
                class="sc-bqyKva pvONl"
              >
                <p
                  class="Discount"
                  data-testid="Discount"
                >
                  USD 0.00
                </p>
              </div>
            </div>
            <div
              class="cart-row  total"
            >
              <p>
                Total
              </p>
              <div
                class="sc-bqyKva pvONl"
              >
                <p
                  class="Total"
                  data-testid="Total"
                >
                  USD 0.00
                </p>
              </div>
            </div>
          </div>
        </div>
        <button
          class="checkout-button"
          data-testid="submit-order"
          disabled=""
          type="button"
        >
          <span>
            CHECKOUT
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
